libs/isograph-react/src/core/garbageCollection.ts

```diff
@@ -51,21 +52,40 @@ export function garbageCollectEnvironment(environment: IsographEnvironment) {
     recordReachableIds(environment.store, query, retainedIds);
   }
 
-  for (const dataId in environment.store) {
-    if (!retainedIds.has(dataId)) {
-      delete environment.store[dataId];
+  for (const typeName in environment.store) {
+    const dataById = environment.store[typeName];
+    if (dataById == null) continue;
+
+    // delete all objects
+    if (!retainedIds[typeName]) {
+      delete environment.store[typeName];
+      continue;
+    }
+
+    for (const dataId in dataById) {
+      if (!retainedIds[typeName].has(dataId)) {
```

Comment:
let's store `retainedIds[typename]` in a local variable, so that we don't need to do the lookup on every loop iteration